### ISSUE TITLE
Server vs Client storage

### ISSUE THREAD

**Issue by chrisgraham** [hybridauth/hybridauth#1201]
I think Hybridauth is close to perfect, but I don't think the storage model is right.

The overarching issue is that you cannot rely on PHP sessions to be stable. session.gc_maxlifetime defaults to 1 day, and could be longer or shorter depending on configuration of any site on the same server (basically). And you shouldn't rely on a session to be long lived as by design sessions are meant to be short-lived.

The idiom for OAuth key storage should be a long-lived one. Someone logging into a website with OAuth expects that to persist over a time span of months, not hours.

Fortunately Hybridauth doesn't bake in session storage, as we can provide our own storage manager.

It's normal to store this longer term stuff in regular cookies. I know this is not ideal because of the overhead of that, but storing this stuff server-side is just not reliable when you consider session data being unpredictably being flushed out of servers or not synched between multiple servers for many use cases. Better to stick to conventions IMO and then hope those conventions get better in the future.

However, there is a baked in assumption - that data storage will still be server-side because 'authorization_state' is used for security and should therefore not be stored on client.

I suggest adding a parameter to the storeData method to indicate whether storage must be server side. I think it only would need flagging for 'authorization_state'.

I can make a pull request, if we're on the same page. Ideally I'd make it with a Cookie storage manager built-in that people can choose.

**Comment by nightw0lv** [comment]
you have valid points, and i believe too that hybridauth is only maintenance on api changes that could mean should improve the data storage at this point

**Comment by chrisgraham** [comment]
Ok I've looked at this more...

It is the case that the Facebook JS SDK handles logins via cookies, but this does not have a refresh token.
The whole point of refresh tokens are that they should not ever be directly held by the client because they should not be capturable.

Plus also Hybridauth has the "<provider>.user" values which need to be permanently held but should not be tamperable, thus they need to be server side.

So my proposal really isn't going to work without further differentiation, enough differentiation to make it too messy IMO.

And also actually PHP does allow its sessions to not use session cookies via the session.cookie_lifetime option. So I'm just going to use this allow with session_name to firewall it from anything else that might mess with the session config and conflict. Hybridauth can be in its own PHP permanent session. People with multiple servers can still make their own backends that implement sessions in some other way.

Closing this issue.
